### PR TITLE
Add FusedRMSNorm (Triton kernel, +15% eager), Add NPLayerNorm, Enable config selectable Norm Type

### PR DESCRIPTION
**Pull request by lessw2020**
This PR has multiple aspects:
1 - Adds a new Triton based Fused RMSNorm I wrote.  I've verified it's numerical accuracy on both forward and backward with a unit test. 
It improves MFU by +15% with FSDP2 7B, and compiled slightly by +1.2%:
Screenshot 2024-03-29 at 5 18 14 PM

2 - Adds norms.py to house all 4 norm types, and standardizes to [layernorm / np_layernorm / rmsnorm / fused_rmsnorm].  Norms.py has a create_norms function that then creates the appropriate norm. 

3 - Adds np_layernorm, which is layernorm with no affine transformation. 

4 - Updates model.py to now support plug and play of any supported norm. 

Thus instead of this type of if/then logic in the model class:
Screenshot 2024-03-30 at 1 52 07 PM

We simply have this:
Screenshot 2024-03-30 at 1 52 23 PM

This then allows for easy plug and play of any norm type with no fiddling around in the model code. 

5 - updates run_llama_train.sh to randomly select a port vs previous fixed port number. (thanks @yifuwang for this tip!)

6 - Now users can quickly select the norm of their choice via the config file:
Screenshot 2024-03-30 at 3 01 43 PM

7 - adds a NotImpl error if users try to run TP + fused_rnsmorm to avoid any confusion (per @tianyu-l feedback):
~~~
NotImplementedError: fused_rmsnorm not yet compatible with TP. Please use rmsnorm.
~~~

### CODE REVIEW

File: torchtrain/modules/norms/fused_rms_norm.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        triton.Config({}, num_warps=2),
+        triton.Config({}, num_warps=4),
+        triton.Config({}, num_warps=8),
+        triton.Config({}, num_warps=16),
+        triton.Config({}, num_warps=32),
+    ],
+    key=["N"],
+)
+@triton.jit
+def _rms_norm_fwd_kernel(
```
Comment (wanchaol) [inline torchtrain/modules/norms/fused_rms_norm.py]: IMO we should also register the fwd/bwd rmsnorm kernel as a PyTorch op, this is so that:
1. making it a custom op makes it compatible with PT2, which I believe it's currently graph breaking on the FusedRMSNorm path if we turn on torch.compile
2. it allows other components (i.e. DTensor) to provide sharding rule to this custom op so that it would compatible with the tensor parallelism

File: torchtrain/modules/norms/fused_rms_norm.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (wanchaol) [inline torchtrain/modules/norms/fused_rms_norm.py]: I think we should have a single file `norm/normalization.py` which encapsulates:
1. FusedRMSNorm kernel as it's already in this file
2. NPLayerNorm
3. RMSNormPython

Then we should also just have a simple function in this file to create each individual normalization layer, i.e.:
`def create/build_norm(norm_type: str, *args, *kwargs)`

where inside this function we just simply check which norm type it is and then create the corresponding layer. Let's not complicate things with `NormBase/NormType` unless we absolutely need to.

We can simply put this file in the `models` folder as this is the first case we want to switch between different impls, once we have more other use cases, we can create a separate `modules` folder.

File: torchtrain/parallelisms/parallelize_llama.py [hunk 3]
```diff
@@ -221,7 +221,7 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
                 )
             # As an optimization, do not reshard after forward for the last
             # transformer block since FSDP would prefetch it immediately
-            reshard_after_forward = layer_id < len(model.layers) - 1
+            reshard_after_forward = False  # layer_id < len(model.layers) - 1
```
Comment (wanchaol) [inline torchtrain/parallelisms/parallelize_llama.py]: why we need to turn this to False?

Comment (awgu) [reply torchtrain/parallelisms/parallelize_llama.py]: I am guessing this is not meant to be committed and is for MFU testing.

Comment (lessw2020) [reply torchtrain/parallelisms/parallelize_llama.py]: ah @awgu  is correct - this was not meant to stay in the commit.  
I'll remove - I was just trying to compare MFU numbers.
Thanks for flagging @wanchaol

File: torchtrain/modules/norms/fused_rms_norm.py (hunk 2 above)
Comment (lessw2020) [reply torchtrain/modules/norms/fused_rms_norm.py]: I've merged everything into a single norms.py file and moved this under models dir. 
added a create_norm function which handles norm layer creation and removed the StrEnum/NormType.
I left the NormBase as otherwise we end up repeating a lot of code in each Normclass which violates the DRY principle...I think it's pretty simple to see the relationship between base and each impl and makes the code shorter/simpler.

File: .flake8 [hunk 4]
```diff
@@ -8,7 +8,7 @@ max-line-length = 120
 # N817 ignored because importing using acronyms is convention (DistributedDataParallel as DDP)
 # E731 allow usage of assigning lambda expressions
 ignore =
-    E203,E305,E402,E501,E721,E741,F405,F821,F841,F999,W503,W504,C408,E302,W291,E303,N812,N817,E731
+    E203,E305,E402,E501,E721,E741,F405,F821,F841,F999,W503,W504,C408,E302,W291,E303,N812,N817,E731,N803,N806
```
Comment (tianyu-l) [inline .flake8:12]: Is this meant to be in the commit? If so, could you also add explanations for them above?

File: run_llama_train.sh [hunk 5]
```diff
@@ -19,6 +19,6 @@ LOG_RANK=${LOG_RANK:-0}
 
 CONFIG_FILE=${CONFIG_FILE:-"./train_configs/debug_model.toml"}
 
-torchrun --nproc_per_node=${NGPU} --rdzv_endpoint="localhost:5972" \
+torchrun --nproc_per_node=${NGPU} --rdzv_endpoint="localhost:5975" \
```
Comment (tianyu-l) [inline run_llama_train.sh]: ditto: how about this one?

File: torchtrain/config_manager.py [hunk 6]
```diff
@@ -117,6 +117,12 @@ def __init__(self):
             default="debugmodel",
             help="which model config to train",
         )
+        self.parser.add_argument(
+            "--model.norm_type",
+            type=str,
+            default="rms",
```
Comment (tianyu-l) [inline torchtrain/config_manager.py]: Currently there are different defaults spread everywhere, sometimes `"rms"` sometimes `"fusedrms"`, making it hard to understand the code. Can we make this cleaner, or have some sort of guideline?

File: torchtrain/models/norms.py [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+        if self.weight is not None:
+            torch.nn.init.ones_(self.weight)  # type: ignore
+
+    def reset_parameters(self):
+        if self.weight is not None:
+            torch.nn.init.ones_(self.weight)  # type: ignore
+
+
+class LayerNorm(NormBase):
+    """Classical LayerNorm, without bias."""
```
Comment (tianyu-l) [inline torchtrain/models/norms.py]: 1. `nn.LayerNorm` by default enables bias, do we want to have something different?
2. Can we just use `nn.LayerNorm` in this case and also `NPLayerNorm`?

If so these two cases become simpler, and we may not need the `NormBase` for the other two quite different cases.

File: torchtrain/utils.py [hunk 8]
```diff
@@ -161,3 +162,23 @@ class Style:
     dim = "\033[2m"
     normal = "\033[22m"
     reset = "\033[0m"
+
+
+class StrEnum(str, enum.Enum):
```
Comment (tianyu-l) [inline torchtrain/utils.py]: Can we remove this?

File: train_configs/llama_7b.toml [hunk 9]
```diff
@@ -4,14 +4,14 @@ dump_folder = "./outputs"
 description = "LLaMA 7B training"
 
 [profiling]
-run_profiler = true
+run_profiler = false
```
Comment (tianyu-l) [inline train_configs/llama_7b.toml]: pls change experimental settings back :)

File: .flake8 (hunk 4 above)
Comment (lessw2020) [reply .flake8:12]: yes, this is intentional - these two errors (N803 and N806) relate to not having caps in function param args.  For python this makes sense...however, for Triton code, there is default caps for incoming params to differentiate from internal variables.  (i.e. X, M, N are all common default use params in Triton).
thus there is a direct conflict between Python vs Triton coding styles.
I am not able to suppress the errors by appending the usual # noqa: N803 after the function declarations, I believe because the function is wrapped with @triton.jit.
Thus I added this to avoid this issue so that the rest of the Triton code can still be checked.

File: torchtrain/config_manager.py (hunk 6 above)
Comment (lessw2020) [reply torchtrain/config_manager.py]: I'm not following when you say 'spread everywhere'.  There's just this single default here, and then the default in debug_model.toml and 7b_toml. 
rms would be rmsnorm and fusedrms would be fusedrms, the fused version.
Originally I made an enum to enforce consistent naming, but Wanchao wanted to remove it.  Thus, since we are using a str now, I've added a generic catch all to each layernorm in the create_norm function so that having an _, (i.e. rms_norm) or caps (RmsNorm) all map to the same thing in case users don't follow things exactly.

File: torchtrain/models/norms.py (hunk 7 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: 1 - we do not want to have default bias in LayerNorm and that's why it's not default here.  bias has almost universally been dropped in training for the past two years as it just adds compute with no net gain.
2 - haivng our own class means we can better update the weights if we want to future add normalized shapes, plus adding the new init_weights requirement is much easier with our own class. 
3 - all norms share some common elements - eps, reset_weights, init_weights so I prefer to have the base class handle that and the norms then implement the forward function.

File: train_configs/llama_7b.toml (hunk 9 above)
Comment (lessw2020) [reply train_configs/llama_7b.toml]: done, thanks for flagging.

File: run_llama_train.sh (hunk 5 above)
Comment (lessw2020) [reply run_llama_train.sh]: this was just b/c that port was locked up on my machine and unusable.  I don't think it matters which port number is our default here, but I can change it back if it's an issue.

File: torchtrain/config_manager.py (hunk 6 above)
Comment (tianyu-l) [reply torchtrain/config_manager.py]: Here are some default values:
1. rms in config_manager.py
2. fusedrms in ModelArgs
3. fusedrms in debug_model.toml
4. rms in other toml files

Since the overwritten order is 3,4 > 1 > 2, so the concrete questions are:
A. Can we change fusedrms to rms in ModelArgs to unify 1 and 2?
B. Why do we have different norm_type in different toml files, can we change them all to rms?

File: .flake8 (hunk 4 above)
Comment (tianyu-l) [reply .flake8:12]: Sounds great! Please add some comments to the line above.

File: run_llama_train.sh (hunk 5 above)
Comment (tianyu-l) [reply run_llama_train.sh]: I see. Is the port still locked? If so we can keep this change to be convenient for you; if not please change it back just to have minimal change needed.

File: torchtrain/config_manager.py (hunk 6 above)
Comment (tianyu-l) [reply torchtrain/config_manager.py]: Also in offline discussion, we were discussing fusedrms doesn't work with TP right now. Shall we note it somewhere (e.g. in config_manager.py and/or toml files) so people don't get frustrated when trying?

File: torchtrain/models/norms.py [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+
+    Returns:
+        The created normalization layer.
+
+    Raises:
+        NotImplementedError: If an unknown norm_type is provided.
+    """
+    norm_type = norm_type.lower()  # Normalize to lowercase
+
+    if norm_type in ("layernorm", "layer_norm"):
```
Comment (awgu) [inline torchtrain/models/norms.py]: nit: Should we just canonicalize and not support different variations of the same name (with different `_` usages)?

I worry this will create more confusion than desired. I imagine the user should be able to query somehow what the supported norms are, and it would be a lot of info if we get >1 string representing each actual type.

Comment (awgu) [reply torchtrain/models/norms.py]: If we really want to keep this, maybe we should have a function to canonicalize the name (i.e. in this case, remove `_`).

File: torchtrain/models/norms.py [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+        super().__init__()
+
+        self.eps = eps
+        self.normalized_shape = (size,)
+        if elementwise_affine:
+            self.weight = nn.Parameter(
+                torch.ones(
+                    self.normalized_shape,
+                )
+            )
```
Comment (awgu) [inline torchtrain/models/norms.py]: nit: This is stylistic, but there is a lot of verticality that might not be needed haha
```suggestion
            self.weight = nn.Parameter(torch.ones(self.normalized_shape))
```
(same with a few other places)

File: torchtrain/models/norms.py (hunk 7 above)
Comment (awgu) [reply torchtrain/models/norms.py]: I think it would be helpful to include some explanation in the code for why we have a custom `LayerNorm` instead of using `nn.LayerNorm`.

> if we want to future add normalized shapes

I was curious what this means.

File: torchtrain/models/norms.py [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+
+
+class NPLayerNorm(NormBase):
+    """Non Parametric LayerNorm - no affine transform."""
+
+    def __init__(
+        self,
+        dim: int,
+        eps: float = 1e-6,
+        elementwise_affine: Optional[bool] = False,
```
Comment (awgu) [inline torchtrain/models/norms.py]: Why is this an argument if `NPLayerNorm` should always default to `False`? Should we remove this argument?

File: torchtrain/models/norms.py [hunk 13]
```diff
@@ -0,0 +1,409 @@
...
+        eps (float): A small value added to the denominator for numerical stability.
+        weight (nn.Parameter): Learnable scaling parameter.
+
+    """
+
+    def __init__(self, dim: int, eps: float = 1e-6):
+        super().__init__(size=dim, eps=eps)
+
+    def _norm(self, x: torch.Tensor):
+        """
```
Comment (awgu) [inline torchtrain/models/norms.py]: I know you are copying this over, but I think we can remove these comments. Personally, they do not add much value, especially when they are not going to be publicly rendered.

File: torchtrain/models/norms.py [hunk 14]
```diff
@@ -0,0 +1,409 @@
...
+    row = tl.program_id(0)
+    cols = tl.arange(0, block_N)
+
+    # Load input data and weights
+    mask = cols < N
+    x = tl.load(X + row * stride_x + cols, mask=mask, other=0.0).to(tl.float32)
+    w = tl.load(W + cols, mask=mask, other=0.0).to(tl.float32)
+
+    # Compute mean and variance
+    # xbar = tl.sum(x, axis=0) / tl.max(tl.sum(mask, axis=0), 1)
```
Comment (awgu) [inline torchtrain/models/norms.py]: Is this meant to be kept in as a comment?

File: torchtrain/models/norms.py [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+            M,
+            N,
+            block_N,
+        )
+
+        ctx.eps = eps
+        ctx.save_for_backward(x, weight, rstd)
+        ctx.x_shape_start = x_shape_start
+
+        y = y.reshape(x_shape_start)
```
Comment (awgu) [inline torchtrain/models/norms.py:254]: Do you know if this can be a `view()` instead of `reshape`?

File: torchtrain/models/norms.py [hunk 16]
```diff
@@ -0,0 +1,409 @@
...
+        return y
+
+    @staticmethod
+    def backward(ctx, dy):
+        x, weight, rstd = ctx.saved_tensors
+        eps = ctx.eps
+        x_shape_start = ctx.x_shape_start
+
+        # Flatten input and output gradients
+        dy = dy.reshape(-1, dy.shape[-1])
```
Comment (awgu) [inline torchtrain/models/norms.py]: ditto: can this be `view`?

File: torchtrain/models/norms.py [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+            rstd,
+            _dw,
+            eps,
+            M,
+            N,
+            rows_per_sm,
+            block_N,
+        )
+        dw = _dw.sum(0).to(weight.dtype)
+        dx = dx.reshape(x_shape_start)
```
Comment (awgu) [inline torchtrain/models/norms.py]: ditto

File: torchtrain/utils.py (hunk 8 above)
Comment (awgu) [reply torchtrain/utils.py]: I agree that we prefer to not include this in torchtrain.

File: torchtrain/models/norms.py [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+        "fused_rms_layernorm",
+        "fusedrms",
+        "fusedrmsnorm",
+    ):
+        return FusedRMSNorm(dim, eps=eps)
+    else:
+        raise NotImplementedError(f"Unknown norm_type: '{norm_type}'")
+
+
+class NormBase(nn.Module):
```
Comment (wanchaol) [inline torchtrain/models/norms.py]: Oh I think we don't need to have a `NormBase` anymore since `create_norm` would just handle all the cases we supported. In particular I think we don't want to re-implement the PyTorch's `nn.LayerNorm`

File: torchtrain/models/norms.py (hunk 10 above)
Comment (wanchaol) [reply torchtrain/models/norms.py]: agreed, let's use consistent naming: 1. rmsnorm 2. fused_rmsnorm 3. layernorm 4. np_layernorm

Then we error our for all other invalid inputs

File: torchtrain/config_manager.py [hunk 19]
```diff
@@ -117,6 +117,12 @@ def __init__(self):
...
+            help="Layer Normalization type to use [layernorm, np_layernorm, rms, fused_rms]",
```
Comment (wanchaol) [inline torchtrain/config_manager.py]: similar comment about the naming, let's just accept these four names: 

1. rmsnorm 2. fused_rmsnorm 3. layernorm 4. np_layernorm

File: run_llama_train.sh (hunk 5 above)
Comment (yifuwang) [reply run_llama_train.sh]: IIRC `--rdzv_backend c10d --rdzv-endpoint=localhost:0` gives you random port everytime.

Comment (lessw2020) [reply run_llama_train.sh]: @yifuwang  - nice, thanks for the tip!  (just tested, works great).

File: torchtrain/config_manager.py (hunk 6 above)
Comment (lessw2020) [reply torchtrain/config_manager.py]: the logic here is that debug_toml and llama_7b will generically run in 1D FSDP eager.  Therefore, since this is likely user first stop to working with torchtrain, we want to put our best foot forward and use fusedrms as default for max speed.

For larger models (70b etc) then its more likely they will use 2D and as you note above, TP doesn't work with fused which is why I made the defaults rms for those.

Comment (lessw2020) [reply torchtrain/config_manager.py]: but yes I will update the ModelArgs to be rms and standardize it a bit more.

File: torchtrain/utils.py (hunk 8 above)
Comment (lessw2020) [reply torchtrain/utils.py]: ok, removed.

File: torchtrain/models/norms.py (hunk 14 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: removed.

File: torchtrain/models/norms.py (hunk 15 above)
Comment (lessw2020) [reply torchtrain/models/norms.py:254]: yes it works fine, might be slightly faster as well.

File: torchtrain/models/norms.py (hunk 16 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: yes, updated. thanks for flagging!

File: torchtrain/config_manager.py (hunk 19 above)
Comment (lessw2020) [reply torchtrain/config_manager.py]: updated and standardized on those 4 now.

File: torchtrain/models/norms.py (hunk 13 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: removed!

File: torchtrain/models/norms.py (hunk 18 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: removed.

File: torchtrain/models/norms.py (hunk 12 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: removed - and per Wanchao, just removed normbase entirely.

File: torchtrain/models/norms.py (hunk 11 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: I didn't actually do this lol (though looks elegant), something triggered the formatter to do it, but I've forced it back into a single line.

File: torchtrain/models/norms.py (hunk 10 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: standardized to those 4 names only and removed the other handling completely.

File: torchtrain/config_manager.py (hunk 6 above)
Comment (lessw2020) [reply torchtrain/config_manager.py]: I added a check during TP parallelization to ensure they are not using fused_rmsnorm, and will advise them to use rmsnorm for now:
~~~
NotImplementedError: fused_rmsnorm not yet compatible with TP. Please use rmsnorm.
~~~

File: torchtrain/modules/norms/fused_rms_norm.py (hunk 1 above)
Comment (lessw2020) [reply torchtrain/modules/norms/fused_rms_norm.py]: I'd like to split this aspect into it's own new PR since getting the fused kernel landed is high pri. 
Created https://github.com/pytorch/torchtrain/issues/199 to track this.

File: torchtrain/models/norms.py [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+    """Fused RMS Norm, wraps a fused Triton Kernel"""
+
+    def __init__(
+        self,
+        dim: int,
+        eps: float = 1e-6,
+    ):
+        super().__init__()
+        self.eps = eps
+        self.weight = nn.Parameter(
```
Comment (tianyu-l) [inline torchtrain/models/norms.py]: can we make this into one or two lines?

File: torchtrain/models/norms.py [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+    Attributes:
+        eps (float): A small value added to the denominator for numerical stability.
+        weight (nn.Parameter): Learnable scaling parameter.
+
+    """
+
+    def __init__(self, dim: int, eps: float = 1e-6):
+        super().__init__()
+        self.eps = eps
+        self.weight = nn.Parameter(
```
Comment (tianyu-l) [inline torchtrain/models/norms.py]: ditto

File: torchtrain/models/norms.py [hunk 22]
```diff
@@ -0,0 +1,333 @@
...
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """leverages Triton Fused RMS Norm kernel"""
+        return self.fused_rms_norm_fn(
+            x,
+            self.weight,
+            eps=self.eps,
+        )
+
+    def init_weights(self):
```
Comment (tianyu-l) [inline torchtrain/models/norms.py]: IIRC from @awgu's earlier comment, we don't need `init_weights` for norm classes which only inits parameters they own; keeping `reset_parameters` is enough.

File: torchtrain/models/norms.py [hunk 23]
```diff
@@ -0,0 +1,333 @@
...
+        )
+
+    def _norm(self, x: torch.Tensor):
+        return x * torch.rsqrt(x.pow(2).mean(-1, keepdim=True) + self.eps)
+
+    def forward(self, x: torch.Tensor):
+        output = self._norm(x.float()).type_as(x)
+        return output * self.weight
+
+    def init_weights(self):
```
Comment (tianyu-l) [inline torchtrain/models/norms.py]: ditto

File: train_configs/llama_7b.toml (hunk 9 above)
Comment (tianyu-l) [inline train_configs/llama_7b.toml]: need to change default settings back :)

File: torchtrain/parallelisms/parallelize_llama.py [hunk 24]
```diff
@@ -137,6 +137,11 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
         raise NotImplementedError("PP not implemented yet.")
 
     if parallel_dims.tp_enabled:
+        if job_config.model.norm_type == "fused_rmsnorm":
+            raise NotImplementedError(
+                "fused_rmsnorm not yet compatible with TP. Please use rmsnorm."
```
Comment (wanchaol) [inline torchtrain/parallelisms/parallelize_llama.py]:
```suggestion
                "fused_rmsnorm not yet compatible with TP. Please use layernorm or rmsnorm."
```

File: torchtrain/models/norms.py (hunk 20 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: flake8 did the multi-line reformat, but I figured out if I don't have a trailing comma in the function param list, then it leaves it alone during pre-commit, so it stays as single line. 
fixed.

File: torchtrain/models/norms.py (hunk 21 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: fixed

File: train_configs/llama_7b.toml (hunk 9 above)
Comment (lessw2020) [reply train_configs/llama_7b.toml]: fixed.

File: torchtrain/models/norms.py (hunk 22 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: good point - ran a quick test to verify this is correct and have removed the init weights.

File: torchtrain/models/norms.py (hunk 23 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: fixed

File: train_configs/llama_7b.toml [hunk 25]
```diff
@@ -30,15 +31,19 @@ lr = 3e-4
...
 
 [training]
-batch_size = 8
+batch_size = 5
 seq_len = 2048
-warmup_steps = 200  # lr scheduler warm up
+warmup_steps = 20  # lr scheduler warm up
 max_norm = 1.0  # grad norm clipping
-steps = 1000
+steps = 41
```
Comment (wanchaol) [inline train_configs/llama_7b.toml]: I guess these config un-related changes are for testing purpose? Please revert the changes that are not relevant to this PR

File: torchtrain/models/norms.py [hunk 26]
```diff
@@ -0,0 +1,314 @@
...
+        dx = (wdy - x_hat * c1) * rstd
+
+        # Store input gradient
+        tl.store(DX + row * stride_dx + cols, dx, mask=mask)
+
+    # Store weight gradients
+    tl.store(DW + row_block_id * N + cols, dw, mask=mask)
+
+
+class TTRMSNorm(torch.autograd.Function):
```
Comment (wanchaol) [inline torchtrain/models/norms.py]: IMO maybe call it FusedRMSNorm would be better, but up to you.

but as a follow up, I think we should add some numeric parity tests in the test folder to ensure FusedRMSNorm and RMSNorm numeric parity. This would also give user the confidence to use it.

File: train_configs/llama_7b.toml [hunk 27]
```diff
@@ -22,6 +22,7 @@ folder = ""
 [model]
 name = "llama"
 flavor = "7B"
+norm_type = "fused_rmsnorm" # layernorm / np_layernorm / rms / fused_rms
```
Comment (wanchaol) [inline train_configs/llama_7b.toml]: I guess the comment need to be updated?

Comment (lessw2020) [reply train_configs/llama_7b.toml]: updated, thanks for catching!

File: train_configs/llama_7b.toml (hunk 25 above)
Comment (lessw2020) [reply train_configs/llama_7b.toml]: updated /fixed.

File: torchtrain/parallelisms/parallelize_llama.py (hunk 24 above)
Comment (lessw2020) [reply torchtrain/parallelisms/parallelize_llama.py]: updated!

File: torchtrain/models/norms.py (hunk 26 above)
Comment (lessw2020) [reply torchtrain/models/norms.py]: updated - went with TritonFusedRMSNorm as FusedRMSNorm is already used earlier, as the exposed wrapper class.
